### PR TITLE
Change Agentd connection Behavior

### PR DESCRIPTION
**Pull request by nicolaspapp**
|Related issue|
|---|
| #5255 |

## Description
 The current schema for connection does not support parametrization and the default times for connection timeouts and reconnection waits results insufficient for a 2 server configuration where the second server works as a fallback. This results into agentd daemon turning into fallback server on small disconnection periods of the main server. 

## Configuration options

Two new parameters have been added to the server configuration:
- max_retries
- retry_interval

Example:

```
<client>
  <server>
    <address>server1 </address>
    <port>1514</port>
    <protocol>udp </protocol>
​
     <!-- New options-->
    <max_retries>5</max_retries>
    <retry_interval>5</retry_interval>
  </server>

   ....
</client>
```

Changes description:
- UDP socket has been changed from non-blocking and is now blocking with a timeout.
- Default configuration for socket recv timeout has been changed from 60 seconds to 5 seconds

## Tests

- Compilation without warnings in every supported platform
  - [X] Linux
  - [x] Windows
  - [x] MAC OS X
  - [x] Solaris 10
- [X] Source installation
- [x] Review logs syntax and correct language
- [x] QA templates contemplate the added capabilities

- Memory tests for Linux
  - [x] Scan-build report
  - [x] Coverity
  - [x] Valgrind (memcheck and descriptor leaks check)
- Memory tests for Windows
  - [ ] Scan-build report
  - [ ] Coverity
  - [ ] Dr. Memory
- Memory tests for macOS
  - [ ] Scan-build report
  - [ ] Leaks
  - [ ] AddressSanitizer
- [X] Configuration on demand reports new parameters
- [X] The data flow works as expected (agent-manager-api-app)
- [X] Added unit tests (for new features)
- [x] Manual and integration tests

Integration tests:
- [ ] Run and pass tests defined in [#wazuh-qa/711](https://github.com/wazuh/wazuh-qa/issues/711) for a Linux agent
- [ ] Run and pass tests defined in [#wazuh-qa/711](https://github.com/wazuh/wazuh-qa/issues/711) for a Windows agent

### CODE REVIEW

File: etc/internal_options.conf [hunk 1]
```diff
@@ -311,7 +311,7 @@ agent.min_eps=50
 agent.state_interval=5
 
 # Maximum time waiting for a server response in TCP (seconds) [1..600]
-agent.recv_timeout=60
+agent.recv_timeout=5
```
Comment (vikman90) [inline etc/internal_options.conf:314]: This timeout is very low. I think it should be 60 as pre-defined.

File: src/config/client-config.h [hunk 2]
```diff
@@ -22,6 +22,8 @@ typedef struct agent_server {
     char * rip;
     int port;
     int protocol;
+    int max_retries; ///> Maximum number of connection retries.
```
Comment (vikman90) [inline src/config/client-config.h:25]: The correct Doxygen syntax is:
```
///<
```
(less than symbol).

File: src/config/client-config.c [hunk 3]
```diff
@@ -218,12 +221,17 @@ int Read_Client_Server(XML_NODE node, agent * logr)
     const char *xml_client_addr = "address";
     const char *xml_client_port = "port";
     const char *xml_protocol = "protocol";
+    const char *xml_max_retries = "max_retries";
+    const char *xml_retry_interval = "retry_interval";
 
     int j;
     char f_ip[128];
     char * rip = NULL;
+    // default values
```
Comment (vikman90) [inline src/config/client-config.c:230]: Capitalize comments correctly.

File: src/unit_tests/agent.cmake [hunk 4]
```diff
@@ -10,3 +10,5 @@ add_compile_options(-ggdb -O0 -g -coverage -DTEST_AGENT -DENABLE_AUDIT -DINOTIFY
 
 # Set tests dependencies
 set(TEST_DEPS ${WAZUHLIB} ${WAZUHEXT} -lpthread -lcmocka -fprofile-arcs -ftest-coverage)
+
+add_subdirectory(client-agent)
```
Comment (vikman90) [inline src/unit_tests/agent.cmake:14]: Missing final line.

File: src/unit_tests/winagent.cmake [hunk 5]
```diff
@@ -73,3 +73,5 @@ target_link_libraries(DEPENDENCIES_O ${WAZUHLIB} ${WAZUHEXT} ${PTHREAD} ${STATIC
 # Use --start-group and --end-group to handle circular dependencies
 set(TEST_DEPS -Wl,--start-group ${WAZUHLIB} ${WAZUHEXT} DEPENDENCIES_O -Wl,--end-group ${PTHREAD} ${STATIC_CMOCKA} wsock32 wevtapi shlwapi comctl32 advapi32 kernel32 psapi gdi32 iphlpapi ws2_32 crypt32 -fprofile-arcs -ftest-coverage)
 set(TEST_EVENT_DEPS -Wl,--start-group ${WAZUHLIB} ${WAZUHEXT} DEPENDENCIES_O -Wl,--end-group ${PTHREAD} ${STATIC_CMOCKA} wsock32 wevtapi shlwapi comctl32 advapi32 kernel32 psapi gdi32 iphlpapi ws2_32 crypt32 -fprofile-arcs -ftest-coverage)
+
+add_subdirectory(client-agent)
```
Comment (vikman90) [inline src/unit_tests/winagent.cmake:77]: Missing ending line.

File: src/unit_tests/wrappers/client-agent/start_agent.c [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    check_expected(fd);
+    return 0;
+}
+
+ssize_t wrap_recv(int __fd, void *__buf, size_t __n, int __flags) {
+    char* rcv = (char*) mock_ptr_type(char*);
+    int len = strlen(rcv);
+    snprintf(__buf, len+1, "%s", rcv);
+    return len;
+}
```
Comment (vikman90) [inline src/unit_tests/wrappers/client-agent/start_agent.c:28]: Missing ending newline.

File: src/unit_tests/wrappers/client-agent/start_agent.h [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+ */
+#ifndef UNIT_TEST_START_AGENT
+#define UNIT_TEST_START_AGENT
+
+int wrap_closesocket(int fd);
+
+ssize_t wrap_recv(int __fd, void *__buf, size_t __n, int __flags);
+
+
+#endif
```
Comment (vikman90) [inline src/unit_tests/wrappers/client-agent/start_agent.h:17]: Missing ending newline.

File: src/client-agent/start_agent.c [hunk 8]
```diff
@@ -12,234 +12,150 @@
...
+static void send_msg_on_startup(void);
 
-/* Attempt to connect to all configured servers */
-int connect_server(int initial_id)
+/**
+ * @brief Connects to a specified server
+ * @param server_id index of the specified server from agt servers list
+ * @retval true on success
+ * @retval false when failed
+ * */
```
Comment (vikman90) [inline src/client-agent/start_agent.c:40]: We should add a post-condition like:
```c
* @post The remote IP id (rip_id) is set to server_id if and only if this function succeeds.
```

File: src/client-agent/start_agent.c [hunk 9]
```diff
@@ -305,120 +221,58 @@ static void w_agentd_keys_init (void) {
 }
 
 /**
- * @brief Holds the message reception logic for udp
- * 
- * @param msg message to be sent
+ * @brief Holds the message reception logic for udp and tcp
```
Comment (vikman90) [inline src/client-agent/start_agent.c:224]: Use capital letters for UDP and TCP, they are acronyms.

File: src/client-agent/start_agent.c (hunk 8 above)
Comment (nicolaspapp) [reply src/client-agent/start_agent.c:41]: Changed in last commit but for some reason it is still not showing up in here

**Pull request merged**